File: cdm-sources/cdm/event/qualification/functions/Qualify_StockSplit.java
package cdm.event.qualification.functions;

import cdm.base.math.FinancialUnitEnum;
import cdm.base.math.QuantitySchedule;
import cdm.base.math.UnitType;
import cdm.base.math.functions.FilterQuantityByCurrencyExists;
import cdm.base.math.functions.FilterQuantityByFinancialUnit;
import cdm.base.math.metafields.FieldWithMetaNonNegativeQuantitySchedule;
import cdm.event.common.BusinessEvent;
import cdm.event.common.Instruction;
import cdm.event.common.Trade;
import cdm.event.common.TradeState;
import cdm.event.common.metafields.ReferenceWithMetaTradeState;
import cdm.observable.asset.PriceQuantity;
import cdm.observable.asset.PriceSchedule;
import cdm.observable.asset.metafields.FieldWithMetaPriceSchedule;
import cdm.product.template.TradeLot;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.expression.MapperMaths;
import com.rosetta.model.lib.functions.IQualifyFunctionExtension;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import java.math.BigDecimal;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(Qualify_StockSplit.Qualify_StockSplitDefault.class)
public abstract class Qualify_StockSplit implements RosettaFunction,IQualifyFunctionExtension<BusinessEvent> {
	
	// RosettaFunction dependencies
	//
	@Inject protected FilterQuantityByCurrencyExists filterQuantityByCurrencyExists;
	@Inject protected FilterQuantityByFinancialUnit filterQuantityByFinancialUnit;

	/**
	* @param businessEvent 
	* @return is_event 
	*/
	@Override
	public Boolean evaluate(BusinessEvent businessEvent) {
		Boolean is_event = doEvaluate(businessEvent);
		
		return is_event;
	}

	protected abstract Boolean doEvaluate(BusinessEvent businessEvent);

	protected abstract MapperS<? extends ReferenceWithMetaTradeState> beforeTradeState(BusinessEvent businessEvent);

	protected abstract MapperS<? extends TradeState> afterTradeState(BusinessEvent businessEvent);

	protected abstract MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> beforeQuantities(BusinessEvent businessEvent);

	protected abstract MapperS<BigDecimal> beforeNoOfUnits(BusinessEvent businessEvent);

	protected abstract MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> afterQuantities(BusinessEvent businessEvent);

	protected abstract MapperS<BigDecimal> afterNoOfUnits(BusinessEvent businessEvent);

	protected abstract MapperS<BigDecimal> beforeCurrencyAmount(BusinessEvent businessEvent);

	protected abstract MapperS<BigDecimal> afterCurrencyAmount(BusinessEvent businessEvent);

	protected abstract MapperS<BigDecimal> beforePrice(BusinessEvent businessEvent);

	protected abstract MapperS<BigDecimal> afterPrice(BusinessEvent businessEvent);

	protected abstract MapperS<Boolean> currencyAmountUnchanged(BusinessEvent businessEvent);

	protected abstract MapperS<Boolean> noOfUnitsChanged(BusinessEvent businessEvent);

	protected abstract MapperS<Boolean> cashPriceChanged(BusinessEvent businessEvent);

	protected abstract MapperS<Boolean> adjustmentRatioMatches(BusinessEvent businessEvent);

	public static class Qualify_StockSplitDefault extends Qualify_StockSplit {
		@Override
		protected Boolean doEvaluate(BusinessEvent businessEvent) {
			Boolean is_event = null;
			return assignOutput(is_event, businessEvent);
		}
		
		protected Boolean assignOutput(Boolean is_event, BusinessEvent businessEvent) {
			is_event = areEqual(currencyAmountUnchanged(businessEvent), MapperS.of(true), CardinalityOperator.All).and(areEqual(noOfUnitsChanged(businessEvent), MapperS.of(true), CardinalityOperator.All)).and(areEqual(cashPriceChanged(businessEvent), MapperS.of(true), CardinalityOperator.All)).and(areEqual(adjustmentRatioMatches(businessEvent), MapperS.of(true), CardinalityOperator.All)).get();
			
			return is_event;
		}
		
		@Override
		protected MapperS<? extends ReferenceWithMetaTradeState> beforeTradeState(BusinessEvent businessEvent) {
			return MapperS.of(MapperS.of(businessEvent).<Instruction>mapC("getInstruction", _businessEvent -> _businessEvent.getInstruction()).<ReferenceWithMetaTradeState>map("getBefore", instruction -> instruction.getBefore()).get());
		}
		
		@Override
		protected MapperS<? extends TradeState> afterTradeState(BusinessEvent businessEvent) {
			return MapperS.of(MapperS.of(businessEvent).<TradeState>mapC("getAfter", _businessEvent -> _businessEvent.getAfter()).get());
		}
		
		@Override
		protected MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> beforeQuantities(BusinessEvent businessEvent) {
			return MapperS.of(beforeTradeState(businessEvent).<TradeState>map("Type coercion", referenceWithMetaTradeState -> referenceWithMetaTradeState == null ? null : referenceWithMetaTradeState.getValue()).<Trade>map("getTrade", tradeState -> tradeState.getTrade()).<TradeLot>mapC("getTradeLot", trade -> trade.getTradeLot()).get()).<PriceQuantity>mapC("getPriceQuantity", tradeLot -> tradeLot.getPriceQuantity()).<FieldWithMetaNonNegativeQuantitySchedule>mapC("getQuantity", priceQuantity -> priceQuantity.getQuantity());
		}
		
		@Override
		protected MapperS<BigDecimal> beforeNoOfUnits(BusinessEvent businessEvent) {
			return MapperS.of(MapperC.of(filterQuantityByFinancialUnit.evaluate(beforeQuantities(businessEvent).<QuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).getMulti(), FinancialUnitEnum.SHARE)).get()).<BigDecimal>map("getValue", quantitySchedule -> quantitySchedule.getValue());
		}
		
		@Override
		protected MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> afterQuantities(BusinessEvent businessEvent) {
			return afterTradeState(businessEvent).<Trade>map("getTrade", tradeState -> tradeState.getTrade()).<TradeLot>mapC("getTradeLot", trade -> trade.getTradeLot()).<PriceQuantity>mapC("getPriceQuantity", tradeLot -> tradeLot.getPriceQuantity()).<FieldWithMetaNonNegativeQuantitySchedule>mapC("getQuantity", priceQuantity -> priceQuantity.getQuantity());
		}
		
		@Override
		protected MapperS<BigDecimal> afterNoOfUnits(BusinessEvent businessEvent) {
			return MapperS.of(MapperC.of(filterQuantityByFinancialUnit.evaluate(afterQuantities(businessEvent).<QuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).getMulti(), FinancialUnitEnum.SHARE)).get()).<BigDecimal>map("getValue", quantitySchedule -> quantitySchedule.getValue());
		}
		
		@Override
		protected MapperS<BigDecimal> beforeCurrencyAmount(BusinessEvent businessEvent) {
			return MapperS.of(distinct(MapperC.<QuantitySchedule>of(filterQuantityByCurrencyExists.evaluate(beforeQuantities(businessEvent).<QuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).getMulti())).<BigDecimal>map("getValue", quantitySchedule -> quantitySchedule.getValue())).get());
		}
		
		@Override
		protected MapperS<BigDecimal> afterCurrencyAmount(BusinessEvent businessEvent) {
			return MapperS.of(distinct(MapperC.<QuantitySchedule>of(filterQuantityByCurrencyExists.evaluate(afterQuantities(businessEvent).<QuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).getMulti())).<BigDecimal>map("getValue", quantitySchedule -> quantitySchedule.getValue())).get());
		}
		
		@Override
		protected MapperS<BigDecimal> beforePrice(BusinessEvent businessEvent) {
			final MapperC<FieldWithMetaPriceSchedule> thenArg0 = MapperS.of(beforeTradeState(businessEvent).<TradeState>map("Type coercion", referenceWithMetaTradeState -> referenceWithMetaTradeState == null ? null : referenceWithMetaTradeState.getValue()).<Trade>map("getTrade", tradeState -> tradeState.getTrade()).<TradeLot>mapC("getTradeLot", trade -> trade.getTradeLot()).get()).<PriceQuantity>mapC("getPriceQuantity", tradeLot -> tradeLot.getPriceQuantity()).<FieldWithMetaPriceSchedule>mapC("getPrice", priceQuantity -> priceQuantity.getPrice())
				.filterItemNullSafe(item -> areEqual(item.<PriceSchedule>map("Type coercion", fieldWithMetaPriceSchedule -> fieldWithMetaPriceSchedule == null ? null : fieldWithMetaPriceSchedule.getValue()).<UnitType>map("getPerUnitOf", priceSchedule -> priceSchedule.getPerUnitOf()).<FinancialUnitEnum>map("getFinancialUnit", unitType -> unitType.getFinancialUnit()), MapperS.of(FinancialUnitEnum.SHARE), CardinalityOperator.All).get());
			final MapperC<BigDecimal> thenArg1 = thenArg0
				.mapItem(item -> item.<PriceSchedule>map("Type coercion", fieldWithMetaPriceSchedule -> fieldWithMetaPriceSchedule == null ? null : fieldWithMetaPriceSchedule.getValue()).<BigDecimal>map("getValue", priceSchedule -> priceSchedule.getValue()));
			return MapperS.of(thenArg1.get());
		}
		
		@Override
		protected MapperS<BigDecimal> afterPrice(BusinessEvent businessEvent) {
			final MapperC<FieldWithMetaPriceSchedule> thenArg0 = MapperS.of(afterTradeState(businessEvent).<Trade>map("getTrade", tradeState -> tradeState.getTrade()).<TradeLot>mapC("getTradeLot", trade -> trade.getTradeLot()).get()).<PriceQuantity>mapC("getPriceQuantity", tradeLot -> tradeLot.getPriceQuantity()).<FieldWithMetaPriceSchedule>mapC("getPrice", priceQuantity -> priceQuantity.getPrice())
				.filterItemNullSafe(item -> areEqual(item.<PriceSchedule>map("Type coercion", fieldWithMetaPriceSchedule -> fieldWithMetaPriceSchedule == null ? null : fieldWithMetaPriceSchedule.getValue()).<UnitType>map("getPerUnitOf", priceSchedule -> priceSchedule.getPerUnitOf()).<FinancialUnitEnum>map("getFinancialUnit", unitType -> unitType.getFinancialUnit()), MapperS.of(FinancialUnitEnum.SHARE), CardinalityOperator.All).get());
			final MapperC<BigDecimal> thenArg1 = thenArg0
				.mapItem(item -> item.<PriceSchedule>map("Type coercion", fieldWithMetaPriceSchedule -> fieldWithMetaPriceSchedule == null ? null : fieldWithMetaPriceSchedule.getValue()).<BigDecimal>map("getValue", priceSchedule -> priceSchedule.getValue()));
			return MapperS.of(thenArg1.get());
		}
		
		@Override
		protected MapperS<Boolean> currencyAmountUnchanged(BusinessEvent businessEvent) {
			if (exists(beforeCurrencyAmount(businessEvent)).and(exists(afterCurrencyAmount(businessEvent))).getOrDefault(false)) {
				return areEqual(beforeCurrencyAmount(businessEvent), afterCurrencyAmount(businessEvent), CardinalityOperator.All).asMapper();
			}
			return MapperS.of(false);
		}
		
		@Override
		protected MapperS<Boolean> noOfUnitsChanged(BusinessEvent businessEvent) {
			if (exists(beforeNoOfUnits(businessEvent)).and(exists(afterNoOfUnits(businessEvent))).getOrDefault(false)) {
				return notEqual(afterNoOfUnits(businessEvent), beforeNoOfUnits(businessEvent), CardinalityOperator.Any).asMapper();
			}
			return MapperS.of(false);
		}
		
		@Override
		protected MapperS<Boolean> cashPriceChanged(BusinessEvent businessEvent) {
			if (exists(beforePrice(businessEvent)).and(exists(afterPrice(businessEvent))).getOrDefault(false)) {
				return notEqual(beforePrice(businessEvent), afterPrice(businessEvent), CardinalityOperator.Any).asMapper();
			}
			return MapperS.of(false);
		}
		
		@Override
		protected MapperS<Boolean> adjustmentRatioMatches(BusinessEvent businessEvent) {
			if (exists(beforeNoOfUnits(businessEvent)).and(greaterThan(beforeNoOfUnits(businessEvent), MapperS.of(BigDecimal.valueOf(0)), CardinalityOperator.All)).and(exists(afterNoOfUnits(businessEvent))).and(exists(beforePrice(businessEvent))).and(exists(afterPrice(businessEvent))).and(greaterThan(afterPrice(businessEvent), MapperS.of(BigDecimal.valueOf(0)), CardinalityOperator.All)).getOrDefault(false)) {
				return areEqual(MapperMaths.<BigDecimal, BigDecimal, BigDecimal>divide(afterNoOfUnits(businessEvent), beforeNoOfUnits(businessEvent)), MapperMaths.<BigDecimal, BigDecimal, BigDecimal>divide(beforePrice(businessEvent), afterPrice(businessEvent)), CardinalityOperator.All).asMapper();
			}
			return MapperS.of(false);
		}
	}
		
		@Override
		public String getNamePrefix() {
			return "Qualify";
		}
}

File: cdm-sources/src/cdm/event/qualification/functions/Qualify_StockSplit.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.base.math.functions.FilterQuantityByCurrencyExists import FilterQuantityByCurrencyExists
from cdm.event.common.BusinessEvent import BusinessEvent

__all__ = ['Qualify_StockSplit']


@replaceable
def Qualify_StockSplit(businessEvent: BusinessEvent) -> bool:
    """
    The qualification of StockSplit business event based on (i) an unchanged before/after currency amount (ii) the same adjustment ratio applied to the before/after cash price and number of units.
    
    Parameters 
    ----------
    businessEvent : BusinessEvent
    
    Returns
    -------
    is_event : boolean
    
    """
    self = inspect.currentframe()
    
    
    def _then_fn0():
        return all_elements(rosetta_resolve_attr(self, "beforeCurrencyAmount"), "=", rosetta_resolve_attr(self, "afterCurrencyAmount"))
    
    def _else_fn0():
        return False
    
    def _then_fn1():
        return any_elements(rosetta_resolve_attr(self, "afterNoOfUnits"), "<>", rosetta_resolve_attr(self, "beforeNoOfUnits"))
    
    def _else_fn1():
        return False
    
    def _then_fn2():
        return any_elements(rosetta_resolve_attr(self, "beforePrice"), "<>", rosetta_resolve_attr(self, "afterPrice"))
    
    def _else_fn2():
        return False
    
    def _then_fn3():
        return all_elements((rosetta_resolve_attr(self, "afterNoOfUnits") / rosetta_resolve_attr(self, "beforeNoOfUnits")), "=", (rosetta_resolve_attr(self, "beforePrice") / rosetta_resolve_attr(self, "afterPrice")))
    
    def _else_fn3():
        return False
    
    beforeTradeState = get_only_element(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "businessEvent"), "instruction"), "before"))
    afterTradeState = get_only_element(rosetta_resolve_attr(rosetta_resolve_attr(self, "businessEvent"), "after"))
    beforeQuantities = rosetta_resolve_attr(rosetta_resolve_attr(get_only_element(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "beforeTradeState"), "trade"), "tradeLot")), "priceQuantity"), "quantity")
    beforeNoOfUnits = rosetta_resolve_attr(get_only_element(FilterQuantityByFinancialUnit(rosetta_resolve_attr(self, "beforeQuantities"), rosetta_resolve_attr(FinancialUnitEnum, "SHARE"))), "value")
    afterQuantities = rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "afterTradeState"), "trade"), "tradeLot"), "priceQuantity"), "quantity")
    afterNoOfUnits = rosetta_resolve_attr(get_only_element(FilterQuantityByFinancialUnit(rosetta_resolve_attr(self, "afterQuantities"), rosetta_resolve_attr(FinancialUnitEnum, "SHARE"))), "value")
    beforeCurrencyAmount = get_only_element(set(rosetta_resolve_attr(FilterQuantityByCurrencyExists(rosetta_resolve_attr(self, "beforeQuantities")), "value")))
    afterCurrencyAmount = get_only_element(set(rosetta_resolve_attr(FilterQuantityByCurrencyExists(rosetta_resolve_attr(self, "afterQuantities")), "value")))
    beforePrice = (lambda item: get_only_element(item))((lambda item: map(lambda item: rosetta_resolve_attr(self, "value"), item))(rosetta_filter(rosetta_resolve_attr(rosetta_resolve_attr(get_only_element(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "beforeTradeState"), "trade"), "tradeLot")), "priceQuantity"), "price"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "perUnitOf"), "financialUnit"), "=", rosetta_resolve_attr(FinancialUnitEnum, "SHARE")))))
    afterPrice = (lambda item: get_only_element(item))((lambda item: map(lambda item: rosetta_resolve_attr(self, "value"), item))(rosetta_filter(rosetta_resolve_attr(rosetta_resolve_attr(get_only_element(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "afterTradeState"), "trade"), "tradeLot")), "priceQuantity"), "price"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "perUnitOf"), "financialUnit"), "=", rosetta_resolve_attr(FinancialUnitEnum, "SHARE")))))
    currencyAmountUnchanged = if_cond_fn((rosetta_attr_exists(rosetta_resolve_attr(self, "beforeCurrencyAmount")) and rosetta_attr_exists(rosetta_resolve_attr(self, "afterCurrencyAmount"))), _then_fn0, _else_fn0)
    noOfUnitsChanged = if_cond_fn((rosetta_attr_exists(rosetta_resolve_attr(self, "beforeNoOfUnits")) and rosetta_attr_exists(rosetta_resolve_attr(self, "afterNoOfUnits"))), _then_fn1, _else_fn1)
    cashPriceChanged = if_cond_fn((rosetta_attr_exists(rosetta_resolve_attr(self, "beforePrice")) and rosetta_attr_exists(rosetta_resolve_attr(self, "afterPrice"))), _then_fn2, _else_fn2)
    adjustmentRatioMatches = if_cond_fn((((((rosetta_attr_exists(rosetta_resolve_attr(self, "beforeNoOfUnits")) and all_elements(rosetta_resolve_attr(self, "beforeNoOfUnits"), ">", 0)) and rosetta_attr_exists(rosetta_resolve_attr(self, "afterNoOfUnits"))) and rosetta_attr_exists(rosetta_resolve_attr(self, "beforePrice"))) and rosetta_attr_exists(rosetta_resolve_attr(self, "afterPrice"))) and all_elements(rosetta_resolve_attr(self, "afterPrice"), ">", 0)), _then_fn3, _else_fn3)
    is_event =  (((all_elements(rosetta_resolve_attr(self, "currencyAmountUnchanged"), "=", True) and all_elements(rosetta_resolve_attr(self, "noOfUnitsChanged"), "=", True)) and all_elements(rosetta_resolve_attr(self, "cashPriceChanged"), "=", True)) and all_elements(rosetta_resolve_attr(self, "adjustmentRatioMatches"), "=", True))
    
    
    return is_event

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
